File: packages/wasmvm/wasmlib/src/wasmtypes/scint32.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_INT32_LENGTH: usize = 4;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn int32_decode(dec: &mut WasmDecoder) -> i32 {
    dec.vli_decode(32) as i32
}

pub fn int32_encode(enc: &mut WasmEncoder, value: i32)  {
    enc.vli_encode(value as i64);
}

pub fn int32_from_bytes(buf: &[u8]) -> i32 {
    if buf.len() == 0 {
        return 0;
    }
    if buf.len() != SC_INT32_LENGTH {
        panic("invalid Int32 length");
    }
    i32::from_le_bytes(buf.try_into().expect("WTF?"))
}

pub fn int32_to_bytes(value: i32) -> Vec<u8> {
    value.to_le_bytes().to_vec()
}

pub fn int32_to_string(value: i32) -> String {
    value.to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableInt32 {
    proxy: Proxy
}

impl ScImmutableInt32 {
    pub fn new(proxy: Proxy) -> ScImmutableInt32 {
        ScImmutableInt32 { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        int32_to_string(self.value())
    }

    pub fn value(&self) -> i32 {
        int32_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable i32 in host container
pub struct ScMutableInt32 {
    proxy: Proxy
}

impl ScMutableInt32 {
    pub fn new(proxy: Proxy) -> ScMutableInt32 {
        ScMutableInt32 { proxy }
    }

    pub fn delete(&self)  {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: i32) {
        self.proxy.set(&int32_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        int32_to_string(self.value())
    }

    pub fn value(&self) -> i32 {
        int32_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/scint32.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

export const ScInt32Length = 4;

import {panic} from '../sandbox';
import {intFromString, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function int32Decode(dec: WasmDecoder): i32 {
    return dec.vliDecode(32) as i32;
}

export function int32Encode(enc: WasmEncoder, value: i32): void {
    enc.vliEncode(value as i32);
}

export function int32FromBytes(buf: Uint8Array): i32 {
    if (buf.length == 0) {
        return 0;
    }
    if (buf.length != ScInt32Length) {
        panic('invalid Int32 length');
    }
    let ret: i32 = buf[3];
    ret = (ret & 0x80) ? ret - 0x100 : ret;
    ret = (ret << 8) | buf[2];
    ret = (ret << 8) | buf[1];
    ret = (ret << 8) | buf[0];
    return ret;
}

export function int32ToBytes(value: i32): Uint8Array {
    const buf = new Uint8Array(ScInt32Length);
    buf[0] = value as u8;
    buf[1] = (value >> 8) as u8;
    buf[2] = (value >> 16) as u8;
    buf[3] = (value >> 24) as u8;
    return buf;
}

export function int32FromString(value: string): i32 {
    return intFromString(value, 32) as i32;
}

export function int32ToString(value: i32): string {
    return value.toString();
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableInt32 {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return int32ToString(this.value());
    }

    value(): i32 {
        return int32FromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableInt32 extends ScImmutableInt32 {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: i32): void {
        this.proxy.set(int32ToBytes(value));
    }
}
